service/storage_service.cc

```diff
@@ -763,18 +784,48 @@ void storage_service::join_token_ring(int delay) {
         MigrationManager.announceNewKeyspace(TraceKeyspace.definition(), 0, false);
 #endif
 
-    // At this point our local tokens are chosen (_bootstrap_tokens) and will not be changed unless we bootstrap again.
-    db::system_keyspace::set_bootstrap_state(db::system_keyspace::bootstrap_state::COMPLETED).get();
-
     slogger.debug("Setting tokens to {}", _bootstrap_tokens);
     // This node must know about its chosen tokens before other nodes do
     // since they may start sending writes to this node after it gossips status = NORMAL.
     // Therefore, in case we haven't updated _token_metadata with our tokens yet, do it now.
     _token_metadata.update_normal_tokens(_bootstrap_tokens, get_broadcast_address());
 
+    if (!db::system_keyspace::bootstrap_complete()) {
+        // If we're not bootstrapping, then we shouldn't have chosen a CDC streams timestamp yet.
+        assert(should_bootstrap() || !_cdc_streams_ts);
+    }
+
+    if (!_cdc_streams_ts && db().local().get_config().check_experimental(db::experimental_features_t::CDC)) {
+        // If we didn't choose a CDC streams timestamp at this point, then either
+        // 1. we already bootstrapped, but are upgrading from a non-CDC version (or previous version of CDC),
+        // 2. or we didn't, but we're skipping the streaming phase (seed node/auto_bootstrap=off)
+        //    and directly joining the token ring.
+
+        // In the first case (updating), we'll propose the first generation
+        // if we conclude that we're the node responsible for this.
+        // In the second case we'll always propose, just as normally bootstrapping nodes do.
+
+        if (!db::system_keyspace::bootstrap_complete() /* second case */
+                || cdc::should_propose_first_generation(get_broadcast_address(), _gossiper)) {
+
+            _cdc_streams_ts = cdc::make_new_cdc_generation(
+                    _bootstrap_tokens, _token_metadata, _sys_dist_ks.local(), get_ring_delay(), _for_testing);
+        }
+    }
```

Comment:
Again let's move the body into the cdc module.

Reply:
Copy-pasting my response to Nadav and Asias:

IMO this particular piece of code should be inlined here in join_token_ring, because to understand what it does and why it's here, you need to understand join_token_ring (and consequently, prepare_to_join and init_server... you probably need to understand 2/3 of storage_service.cc to understand how nodes start). There's a bunch of "invisible" relationships and invariants that must hold while this particular piece of code runs.

The comments are there to explain these relationships (I wish there were many more such comments in other parts of join_token_ring and prepare_to_join).

If I were to move this code to a different file, I would have to write more explanations ("this procedure must be called exactly from that place or it won't work"). And then to understand it you would have to open both storage_service.cc and the other file in a vertical split and read them consecutively.

Reply:
There are ways to abstract this. You give a gossip-themed name to the relationships/invariants, and then register cdc as a gossip client. For example (no idea if the name is correct or not), call it gossip_client::pre_commit() and allow a derived class to add its states here.

I'm not suggesting to do it now, because it can expand the work in a fracal-like manner. But we should be aware that we're reducing code quality here by entangling cdc with everything else.